src/terms/Hamiltonian.jl

```diff
@@ -2,121 +2,149 @@
 # corresponding to each term
 # This is the "high-level" interface, provided for convenience
 
-struct HamiltonianBlock
+abstract type HamiltonianBlock end
+
+# Generic HamiltonianBlock
+struct GenericHamiltonianBlock <: HamiltonianBlock
     basis::PlaneWaveBasis
     kpoint::Kpoint
 
     # The operators are vectors of RealFourierOperator,
     # not typed because of type invariance issues.
     operators::Vector  # the original list of RealFourierOperator
                        # (as many as there are terms), kept for easier exploration
-    optimized_operators::Vector  # the optimized list of RealFourierOperator, to be used for applying
+    optimized_operators::Vector  # Optimized list of RealFourierOperator, for application
+end
+
+# More optimized HamiltonianBlock for the important case of a DFT Hamiltonian
+struct DftHamiltonianBlock <: HamiltonianBlock
+    basis::PlaneWaveBasis
+    kpoint::Kpoint
+    operators::Vector
+    optimized_operators::Vector
+
+    # Individual operators for easy access
+    fourier_op::FourierMultiplication
+    real_op::RealSpaceMultiplication
+    nonlocal_op::NonlocalOperator
+    divAgrad_op::Union{Nothing,DivAgradOperator}
+
     scratch  # Pre-allocated scratch arrays for fast application
 end
-function HamiltonianBlock(basis, kpt, operators, scratch)
-    HamiltonianBlock(basis, kpt, operators, optimize_operators_(operators), scratch)
+
+function HamiltonianBlock(basis, kpoint, operators, scratch=ham_allocate_scratch_(basis))
+    optimized_operators = optimize_operators_(operators)
+    if length(optimized_operators) in (3, 4)
+        fourier_ops  = filter(o -> o isa FourierMultiplication,   optimized_operators)
+        real_ops     = filter(o -> o isa RealSpaceMultiplication, optimized_operators)
+        nonlocal_ops = filter(o -> o isa NonlocalOperator,        optimized_operators)
+        divAgrid_ops = filter(o -> o isa DivAgradOperator,        optimized_operators)
+        @assert length(divAgrid_ops) < 2
+
+        if (length(fourier_ops) == length(real_ops) == length(nonlocal_ops) == 1)
+            return DftHamiltonianBlock(basis, kpoint, operators, optimized_operators,
+                                       only(fourier_ops), only(real_ops), only(nonlocal_ops),
+                                       something(divAgrid_ops..., Some(nothing)),
```

Comment:
that pattern doesn't exactly roll off the tongue

Reply:
I hope it's better now.